webodf/lib/ops/OdtDocument.js

```diff
@@ -938,6 +979,48 @@ ops.OdtDocument = function OdtDocument(odfCanvas) {
     };
 
     /**
+     * Process steps being inserted into the document. Will emit a steps inserted signal on
+     * behalf of the caller
+     * @param {!{position: !number}} args
+     * @return {undefined}
+     */
+    this.handleStepsInserted = function(args) {
+        stepsTranslator.handleStepsInserted(args);
+        // signal not used in webodf, but 3rd-party (NVivo)
+        eventNotifier.emit(ops.OdtDocument.signalStepsInserted, args);
+    };
+
+    /**
+     * Process steps being removed from the document. Will emit a steps removed signal on
+     * behalf of the caller
+     * @param {!{position: !number}} args
+     * @return {undefined}
+     */
+    this.handleStepsRemoved = function(args) {
+        stepsTranslator.handleStepsRemoved(args);
+        // signal not used in webodf, but 3rd-party (NVivo)
+        eventNotifier.emit(ops.OdtDocument.signalStepsRemoved, args);
+    };
+
+    /**
+     * Emit the signal that the passed cursor moved.
+     * @param {!ops.OdtCursor} cursor
+     * @return {undefined}
+     */
+    this.emitSignalCursorMoved = function(cursor) {
+        eventNotifier.emit(ops.Document.signalCursorMoved, cursor);
```

Comment:
This one and the undo stack modification definitely should be queued up when called via an operation. This one especially is one of the most frequent causes of the bug this patch is trying to fix.

I would instead suggest passing in the events array into function calls that need to generate signals, as there are a limited number of these.

For the specific edge case that is the shadow cursor, perhaps you can inject or expose a specialised callback to be used ONLY by the shadow cursor. This ensures operations don't get access to the these inappropriate methods.